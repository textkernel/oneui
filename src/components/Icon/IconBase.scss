@import '../../themes/oneui/oneui';

.IconBase {
    display: inline-flex;
    align-self: center;
    fill: currentColor;

    &--isPrimary {
        fill: var(--color-primary);
    }

    &__svg {
        position: relative;
        top: 0.125em;
        height: 1em;
        transition: fill var(--transition-duration);
    }

    &--margin {
        @each $side in (top, right, bottom, left) {
            &_#{ $side } {
                margin-#{ $side }: 0.25em;
            }
        }
    }
}
