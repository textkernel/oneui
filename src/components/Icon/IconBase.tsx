import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import styles from './IconBase.scss';
import { Context } from '../../constants';

export interface IconProps extends React.HTMLAttributes<HTMLDivElement> {
    /** The icon context (e.g. brand, primary, bad, good etc. - defaults to brand) */
    context?: Context;
    /** Adds margin between a given side of the icon and other content */
    margin?: 'top' | 'right' | 'bottom' | 'left';
    /** Absolute size for this icon (size in pixels, aspect ratio is 1:1).
     If not defined, icon will scale and align itself with text. */
    size?: number;
    /** If true, width will set to be automatic */
    preserveAspectRatio?: boolean;
    /** Optional icon title */
    title?: string;
}

export interface Props extends IconProps {
    /** The SVG content */
    children: NotEmptyReactNode;
    /** The SVG viewbox */
    viewBox: string;
}

const { block, elem } = bem('IconBase', styles);

const adjustSize = (preserveAspectRatio: boolean, size?: number) => {
    if (!size) {
        if (!preserveAspectRatio) {
            return { width: '1em' };
        }
        return { width: 'auto' };
    }

    const adjustedSize = Math.max(0, size);

    return {
        top: 'auto',
        width: preserveAspectRatio ? 'auto' : adjustedSize,
        height: adjustedSize,
    };
};

export const IconBase = forwardRef<HTMLDivElement, Props>(
    (
        { children, context, margin, size, preserveAspectRatio = false, title, viewBox, ...rest },
        ref
    ) => {
        return (
            <div ref={ref} {...rest} {...block({ margin, context, ...rest })}>
                <svg
                    {...elem('svg')}
                    aria-labelledby={title ? 'title' : null}
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox={viewBox}
                    style={adjustSize(!!preserveAspectRatio, size)}
                    role="img"
                >
                    {!!title && <title>{title}</title>}
                    {children}
                </svg>
            </div>
        );
    }
);

IconBase.displayName = 'IconBase';
