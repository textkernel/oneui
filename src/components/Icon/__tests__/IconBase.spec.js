import React from 'react';
import toJson from 'enzyme-to-json';
import { IconBase } from '../IconBase';

describe('<IconBase> that renders an SVG wrapper with all options included', () => {
    it('should render a default icon', () => {
        const wrapper = shallow(
            <IconBase viewBox="0 0 100 100" title="Icon base" size={15} margin="right" isPrimary>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should correct negative sizes', () => {
        const wrapper = shallow(
            <IconBase viewBox="0 0 100 100" size={-1}>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(wrapper.find('svg')).not.toContain('title');
        expect(wrapper.find('svg').prop('style').height).toBe(0);
        expect(wrapper.find('svg').prop('style').width).toBe(0);
    });
    it('should not apply proportional styles if no size provided', () => {
        const wrapper = shallow(
            <IconBase viewBox="0 0 100 100">
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(wrapper.find('svg').prop('style').width).toBe('1em');
    });
    it('should apply proportional styles if size is provided', () => {
        const mockSize = 48;
        const wrapper = shallow(
            <IconBase viewBox="0 0 100 100" size={mockSize}>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(wrapper.find('svg').prop('style').height).toBe(mockSize);
        expect(wrapper.find('svg').prop('style').width).toBe(mockSize);
    });
    it('should apply hight only to styles if size is provided with preserveAspectRatio', () => {
        const mockSize = 48;
        const wrapper = shallow(
            <IconBase viewBox="0 0 100 100" size={mockSize} preserveAspectRatio>
                <circle cx="50" cy="50" r="50" />
            </IconBase>
        );
        expect(wrapper.find('svg').prop('style').height).toBe(mockSize);
        expect(wrapper.find('svg').prop('style').width).toBe('auto');
    });
});
