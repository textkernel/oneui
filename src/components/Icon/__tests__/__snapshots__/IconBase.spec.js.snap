// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconBase> that renders an SVG wrapper with all options included should render a default icon 1`] = `
<div
  className="IconBase IconBase--margin_right IconBase--isPrimary"
>
  <svg
    aria-labelledby="title"
    className="IconBase__svg"
    role="img"
    style={
      Object {
        "height": 15,
        "top": "auto",
        "width": 15,
      }
    }
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      Icon base
    </title>
    <circle
      cx="50"
      cy="50"
      r="50"
    />
  </svg>
</div>
`;
